cpp/include/cudf/io/datasource.hpp

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <arrow/buffer.h>
+#include <arrow/io/file.h>
+#include <arrow/io/interfaces.h>
+#include <arrow/io/memory.h>
+
+#include <memory>
+
+#include <cudf/utilities/error.hpp>
+
+namespace cudf {
+//! IO interfaces
+namespace io {
+
+/**
+ * @brief Interface class for providing input data to the readers.
+ */
+class datasource {
+ public:
+  /**
+   * @brief Interface class for buffers that the datasource returns to the caller.
+   *
+   * Provides a basic interface to return the data address and size.
+   */
+  class buffer {
+   public:
+    /**
+     * @brief Returns the buffer size in bytes.
+     */
+    virtual size_t size() const = 0;
+
+    /**
+     * @brief Returns the address of the data in the buffer.
+     */
+    virtual const uint8_t* data() const = 0;
+
+    /**
+     * @brief Base class destructor
+     */
+    virtual ~buffer() {}
+  };
+
+  /**
+   * @brief Creates a source from a file path.
+   *
+   * @param[in] filepath Path to the file to use
+   * @param[in] offset Bytes from the start of the file (the default is zero)
+   * @param[in] size Bytes from the offset; use zero for entire file (the default is zero)
+   */
+  static std::unique_ptr<datasource> create(const std::string& filepath,
+                                            size_t offset = 0,
+                                            size_t size   = 0);
+
+  /**
+   * @brief Creates a source from a memory buffer.
+   *
+   * @param[in] data Pointer to the input data host buffer
+   * @param[in] size Size of the buffer in bytes
+   */
+  static std::unique_ptr<datasource> create(const char* data, size_t size);
+
+  /**
+   * @brief Creates a source from a from an Arrow file.
+   *
+   * @param[in] arrow_file RandomAccessFile to which the API calls are forwarded
+   */
+  static std::unique_ptr<datasource> create(
+    std::shared_ptr<arrow::io::RandomAccessFile> arrow_file);
```

Comment:
What's the reason to leave this as a special constructor vs.  implementing a derived class and using the custom/user sink create() overload?

Reply:
@nvdbaranec  Not sure what you mean. Are you saying that we should let the users implement the derived class?

Reply:
No, I meant implementing it ourselves (as part of this PR).  The idea being, we've got 2 generic ways in :  file system and memory-mapped.   Anything beyond that is more like a custom/user implementation.  Feels like Arrow io stuff would be an example of that.

Reply:
I see and I like the reasoning. 
Well, we already have this implemented, class `arrow_io_source` in `datasource.cpp`. With the overload above removed, we would need to move this class declaration into a header so users can instantiate it.
How does this sound to you?
My only concern is that it would break the existing API, but I guess we are still okay to do so.

Reply:
That sounds fine to me.  Your point about breaking existing the existing API is valid though.  Pinging @OlivierNV  and @jrhemstad  for opinions.

Reply:
Also @felipeblazing 
Any concerns with these API changes of the `arrow::io` support?

Reply:
Improvements (changes) to existing APIs are fine and we are not required to maintain backward compatibility within a major release until 1.0. But we should check with @williamBlazing @felipeblazing about this.

Reply:
Discussed with @vuule and this will be discussed and put in a separate PR after 0.14.

Reply:
@felipeblazing , @williamBlazing , any concern with the suggested changes to `arrow::io` `datasource`? 
From the user's perspective the change is that you need to create an `arrow_io_datasource` object and pass it to `sink_info`, instead of passing an `arrow::io::RandomAccessFile` directly.

Reply:
@vuule I dont think it will be a problem. THanks for asking